---
title: "Figure 4B Vignette"
author: "Jianhong Ou"
vignette: >
  %\VignetteIndexEntry{Figure 4B Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  html_document:
  theme: simplex
  toc: true
  toc_float: true
  toc_depth: 4
  fig_caption: true
---

```{r setup, include=FALSE}
suppressPackageStartupMessages({
  library(geomeTriD)
  library(geomeTriD.documentation)
  library(GenomicRanges)
  library(clusterCrit)
  library(grid)
  library(colorRamps)
  library(TxDb.Mmusculus.UCSC.mm10.knownGene)
  library(org.Mm.eg.db)
  library(geometry)
})
knitr::opts_chunk$set(echo = TRUE, out.width = "900px")
```

# Figure 4B

Figure 4B showcases the geomeTriD package, demonstrating how it presents 3D models along with multiple genomic signals mapped onto single-cell 3D structures.

## Load Libraries

```{r}
library(geomeTriD)
library(geomeTriD.documentation)
library(GenomicRanges)
library(colorRamps)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(geometry)
```

## Present single cell 3D structure for HiRES data

This dataset was generated using [HiRES](https://doi.org/10.1126/science.adg3797), an assay that stands for "Hi-C and RNA-seq employed simultaneously." HiRES enables the simultaneous profiling of single-cell Hi-C and RNA-seq data. The resulting Hi-C data can be used to predict 3D genome structures using `Dip-C`, and the corresponding RNA-seq data can be visualized along these structures using the `geomeTriD` package.

```{r}
## load data for HiRES
extdata <- system.file('extdata', 'GSE223917', package = 'geomeTriD.documentation')
HiRES <- readRDS(file.path(extdata, 'HiRES.radial_glias.G1.chrX.rds')) # Dip-C predicted 3D structure
exprs <- readRDS(file.path(extdata, 'expr.radial_glias.G1.chrX.rds'))# RNA-seq data
pairs <- readRDS(file.path(extdata, 'sel.imput.pairs.chrX.rds')) # selected impute pairs

### supperloop
supperloops <- GRanges(c('chrX:50555744-50635321',
                      'chrX:75725458-75764699', # 4933407K13RiK, NR_029443
                      'chrX:103422010-103484957',
                      'chrX:105040854-105117090')) # 5530601H04RiK, NR_015467 and Pbdc1
names(supperloops) <- c('Firre', 'Dxz4', 'Xist/Tsix', 'x75')
supperloops$label <- names(supperloops)
supperloops$col <- 2:5 ## set colors for each element
supperloops$type <- 'gene' ## set it as gene

## plot region
range <- as(seqinfo(TxDb.Mmusculus.UCSC.mm10.knownGene)['chrX'], 'GRanges')

## annotations
genes <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
genes_symbols <- mget(genes$gene_id, org.Mm.egSYMBOL, ifnotfound = NA)
genes$symbols <- sapply(genes_symbols, `[`, i=1)
geneX <- genes[seqnames(genes)=='chrX']
geneX$label <- geneX$symbols
geneX <- geneX[geneX$symbols %in% rownames(exprs)]

## get data for female, must have mat and pat info for chromosome X.
HiRES.mat <- lapply(HiRES, function(.ele) {
  .ele <- .ele[.ele$parental=='(mat)']
  .ele$parental <- NULL
  .ele
})
HiRES.pat <- lapply(HiRES, function(.ele) {
  .ele <- .ele[.ele$parental=='(pat)']
  .ele$parental <- NULL
  .ele
})
## remove the cells without chrX structures
l.mat <- lengths(HiRES.mat)
l.pat <- lengths(HiRES.pat)
k <- l.mat>0 & l.pat>0
HiRES.mat <- HiRES.mat[k]
HiRES.pat <- HiRES.pat[k]
# Make all GRanges object in a list same length by filling with NA
HiRES <- paddingGRangesList(c(HiRES.mat, HiRES.pat))
HiRES.mat.xyzs <- HiRES$xyzs[seq_along(HiRES.mat)]
HiRES.pat.xyzs <- HiRES$xyzs[-seq_along(HiRES.mat)]
## prepare the maternal and paternal GRanges object with x, y, z coordinates.
HiRES.mat <- lapply(HiRES.mat.xyzs, function(.ele){
  gr <- HiRES$gr
  mcols(gr) <- .ele[, c('x', 'y', 'z')]
  gr
})
HiRES.pat <- lapply(HiRES.pat.xyzs, function(.ele){
  gr <- HiRES$gr
  mcols(gr) <- .ele[, c('x', 'y', 'z')]
  gr
})

## backbone color
resolution <- 3
backbone_colors <- matlab.like2(n=resolution*length(HiRES.mat[[1]]))
backbone_bws <- backbone_colors ## defines the color assigned to the chrX allele
backbone_bws[1001:(length(backbone_colors)-1000)] <- 'gray'

## help function to check the volume of the 3D structure
getV <- function(points){
  vol <- convhulln(points, options='Fa')$vol
}

## calculate the Root Mean Square Deviation (RMSD)
RMSD_mat_pat <- mapply(function(mat, pat){
  mat <- as.data.frame(mcols(mat))
  pat <- as.data.frame(mcols(pat))
  mat <- fill_NA(mat)
  pat <- fill_NA(pat)
  pat <- alignCoor(pat, mat) # do alignment first
  ## normalized to its centroid
  mat.center <- colMeans(mat, na.rm = TRUE)
  pat.center <- colMeans(pat, na.rm = TRUE)
  mat <- t(t(mat) - mat.center)
  pat <- t(t(pat) - pat.center)
  mean(sqrt(rowSums((mat - pat)^2)), na.rm = TRUE)
}, HiRES.mat, HiRES.pat)
## data frame for Xist expression level, total expression level and
## the RMSD between mat and pat
XlinkExpr <- data.frame(Xist=exprs['Xist', ], total=colSums(exprs), RMSD=RMSD_mat_pat)
XlinkExpr <- XlinkExpr[order(XlinkExpr$total), ]
## plot the correlation between RMSD and total chrX expression level
fit <- lm(RMSD ~ total, data = XlinkExpr)
plot(XlinkExpr$total, XlinkExpr$RMSD,
     xlab='Total expression level of X-lined gene',
     ylab='RMSD between maternal and paternal')
abline(fit, col = "blue", lwd=2)

widgets <- lapply(c(head(rownames(XlinkExpr), n=2),
                    tail(rownames(XlinkExpr), n=2)), function(cell_id){
    ## expressions in single cell
    exprSig <- geneX
    exprSig$score <- exprs[geneX$symbols, cell_id]
    ## load the 3D structure for maternal and paternal
    mat_cell <- HiRES.mat[[cell_id]]
    mcols(mat_cell) <- fill_NA(as.data.frame(mcols(mat_cell)))
    pat_cell <- HiRES.pat[[cell_id]]
    mcols(pat_cell) <- fill_NA(as.data.frame(mcols(pat_cell)))
    ## check the volumn of the chrX, 
    ## bigger one is Xa
    ## condensed one is Xi
    v_mat <- getV(as.matrix(mcols(mat_cell)))
    v_pat <- getV(as.matrix(mcols(pat_cell)))
    ## add additional information,
    ## Here we use the selected interactions by impute phases for maternal
    mat_only_pairs <- pairs$mat[[cell_id]]
    mat_only_pairs$color <- 'black'
    mat_only_pairs$lwd <- 4
    mat_cell <- view3dStructure(mat_cell,
                                feature.gr=supperloops,
                                lwd.gene = 4,
                                renderer = 'none',
                                region = range,
                                resolution=resolution,
                                genomicSigs = if(v_mat>v_pat) {
                                  list(mat_rna_reads=exprSig, mat_pairs=mat_only_pairs)
                                } else {list(mat_pairs=mat_only_pairs)},
                                signalTransformFun = c,
                                reverseGenomicSigs = FALSE,
                                show_coor=FALSE,
                                lwd.backbone = 0.25,
                                col.backbone = if(v_mat>v_pat) backbone_bws else backbone_colors)
    ## and paternal only interactions by impute phases
    pat_only_pairs <- pairs$pat[[cell_id]]
    pat_only_pairs$color <- 'black'
    pat_only_pairs$lwd <- 4
    pat_cell <- view3dStructure(pat_cell,
                                feature.gr=supperloops,
                                lwd.gene = 4,
                                renderer = 'none',
                                region = range,
                                resolution=resolution,
                                genomicSigs = if(v_mat<=v_pat) {
                                  list(pat_rna_reads=exprSig, pat_pairs=pat_only_pairs)
                                } else {list(pat_pairs=pat_only_pairs)},
                                signalTransformFun = c,
                                reverseGenomicSigs = FALSE,
                                show_coor=FALSE,
                                lwd.backbone = 0.25,
                                col.backbone = if(v_mat>v_pat) backbone_colors else backbone_bws)
    # widget <-showPairs(mat_cell, pat_cell, title = paste(c('mat', 'pat'), cell_id))
    # tempfile <- paste0('cell.', cell_id, '.html')
    # htmlwidgets::saveWidget(widget, file=tempfile, selfcontained = FALSE, libdir = 'js')
    showPairs(mat_cell, pat_cell,
              title = paste(c('mat', 'pat'), cell_id),
              height=NULL)
})
## low expression of X-linked genes
widgets[[1]]
#widgets[[2]]
## high expression of X-linked genes
#widgets[[3]]
widgets[[4]]
```

## Present single cell 3D structure for ChAIR data

Here we present 3D genome models with integrated genomic signals derived from [Tri-omic](https://doi.org/10.1038/s41592-025-02658-7), also known as ChAIR (capture of chromatin Accessibility, Interaction, and RNA profiles). This method enables simultaneous mapping of chromatin accessibility, chromatin interactions, and gene expression. Paired-end tag (PET) data are used to reconstruct 3D genome structures, while the accompanying RNA-seq and ATAC-seq signals are visualized along these models.

```{r}
## load the 3D structure for PET data
stage <- c('G1', 'S', 'G2M')
extdata <- system.file('extdata', 'ChAIR', package = 'geomeTriD.documentation')
ChAIR_PET_m <- import3dg(file.path(extdata, 'PET', paste0('patski.allele.', stage, '_2_m.pairs.chrX.3dg.gz')))
ChAIR_PET_p <- import3dg(file.path(extdata, 'PET', paste0('patski.allele.', stage, '_2_p.pairs.chrX.3dg.gz')))
## get the chrX coordinates for maternal and paternal
ChAIR_PET_m <- lapply(ChAIR_PET_m, function(.ele) {
  seqlevelsStyle(.ele) <- 'UCSC'
  .ele$parental <- NULL
  subsetByOverlaps(.ele, range)
  })
ChAIR_PET_p <- lapply(ChAIR_PET_p, function(.ele) {
  seqlevelsStyle(.ele) <- 'UCSC'
  .ele$parental <- NULL
  subsetByOverlaps(.ele, range)
})

# import the signals for ATAC-seq and RNA-seq for maternal and paternal
M <- lapply(stage, function(cc){ ## maternal
  importGenomicSigs(c(ATAC=file.path(extdata, 'ATAC', paste0('patski.allele.', cc, '_2.M.bw')),
                      RNA=file.path(extdata, 'RNA', paste0('patski.allele.', cc, '_2.M.bw'))),
                      range = range,
                    cols = c('orange', 'red'),
                    format = 'BigWig')
  })
P <- lapply(stage, function(cc){ ## paternal
  importGenomicSigs(c(ATAC=file.path(extdata, 'ATAC', paste0('patski.allele.', cc, '_2.M.bw')),
                      RNA=file.path(extdata, 'RNA', paste0('patski.allele.', cc, '_2.M.bw'))),
                      range = range,
                    cols = c('orange', 'red'),
                    format = 'BigWig')
})

## set backbone color, make it easy to catch the genomic coordinates
resolution <- 3
backbone_colors <- rev(rainbow(n=resolution*length(ChAIR_PET_m[[1]])))
## create threeJsGeometry objects for maternal and paternal cells
M_cells <- mapply(ChAIR_PET_m, M, FUN=function(model, sig){
  cell <- view3dStructure(model[seqnames(model)=='chrX'],
                  feature.gr = supperloops,
                  genomicSigs = sig,
                  signalTransformFun = log1p,
                  lwd.maxGenomicSigs = 8,
                  reverseGenomicSigs = FALSE,
                  renderer = 'none',
                  resolution = resolution,
                  col.backbone = backbone_colors,
                  region = range,
                  show_coor = FALSE)
  backbone <- extractBackbonePositions(cell)
  sl <- createTADGeometries(supperloops, backbone, type = 'sphere', alpha = 0.8)
  c(cell, sl)
}, SIMPLIFY = FALSE)
P_cells <- mapply(ChAIR_PET_p, P, FUN=function(model, sig){
  cell <- view3dStructure(model[seqnames(model)=='chrX'],
                  feature.gr = supperloops,
                  genomicSigs = sig,
                  signalTransformFun = log1p,
                  lwd.maxGenomicSigs = 8,
                  reverseGenomicSigs = FALSE,
                  renderer = 'none',
                  resolution = resolution,
                  col.backbone = backbone_colors,
                  region = range,
                  show_coor = FALSE)
  backbone <- extractBackbonePositions(cell)
  sl <- createTADGeometries(supperloops, backbone, type = 'sphere', alpha = 0.8)
  c(cell, sl)
}, SIMPLIFY = FALSE)

## show the maternal and paternal structures side by side
#showPairs(M_cells[[1]], P_cells[[1]], title=c('mat G1', 'pat G1')) #G1
showPairs(M_cells[[2]], P_cells[[2]], title=c('mat S', 'pat S')) #S
#showPairs(M_cells[[3]], P_cells[[3]], title=c('mat G2M', 'pat G2M')) #G2M

```

## SessionInfo

```{r sessionInfo}
sessionInfo()
```