---
title: "Figure 5 Vignette"
author: "Jianhong Ou"
vignette: >
  %\VignetteIndexEntry{Figure 5 Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  html_document:
  theme: simplex
  toc: true
  toc_float: true
  toc_depth: 4
  fig_caption: true
---

```{r setup, include=FALSE}
suppressPackageStartupMessages({
  library(geomeTriD)
  library(geomeTriD.documentation)
  library(GenomicRanges)
  library(TxDb.Hsapiens.UCSC.hg38.knownGene)
  library(org.Hs.eg.db)
  library(clusterCrit)
  library(grid)
  library(colorRamps)
})
knitr::opts_chunk$set(echo = TRUE)
```

# Figure 5

The Figure 5 is the showcase for `geomeTriD` package to present multiple genomic 
signals along with 3D models for topologically associating domain (TADs). 

## Load Libraries

```{r loadLibrary}
library(geomeTriD)
library(geomeTriD.documentation)
library(GenomicRanges)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
library(clusterCrit)
library(colorRamps)
```

## Load data and annotations
The data were from [diffraction-limited experiments for IMR90 cells](https://doi.org/10.1126/science.aau1783).

```{r}
## xyz is from diffraction-limited experiments but not the actual xyz cords of STORM experiments.
xyz <- read.csv('https://github.com/BogdanBintu/ChromatinImaging/raw/refs/heads/master/Data/IMR90_chr21-28-30Mb.csv', skip = 1)
## split the XYZ for each chromosome
xyz.s <- split(xyz[, -1], xyz$Chromosome.index)
head(xyz.s[[1]])

range <- GRanges("chr21:28000001-29200000")
### get feature annotations
feature.gr <- getFeatureGR(txdb = TxDb.Hsapiens.UCSC.hg38.knownGene,
                           org = org.Hs.eg.db,
                           range = range)

### get genomic signals, we will use called peaks
colorSets <- c(CTCF="cyan",YY1="yellow",
               RAD21="red", SMC1A="green", SMC3="blue")
prefix <- 'https://www.encodeproject.org/files/'
imr90_url <- c(ctcf = 'ENCFF203SRF/@@download/ENCFF203SRF.bed.gz',
               smc3 = 'ENCFF770ISZ/@@download/ENCFF770ISZ.bed.gz',
               rad21 = 'ENCFF195CYT/@@download/ENCFF195CYT.bed.gz')
imr90_urls <- paste0(prefix, imr90_url)
names(imr90_urls) <- names(imr90_url)
# download the files and import the signals
imr90_genomicSigs <- importSignalFromUrl(imr90_urls, range = range,
                                         cols = list(
                                           'ctcf'=c('#111100', 'cyan'),
                                           'smc3'=c('#000011', 'blue'),
                                           'rad21'=c('#110000', 'red')),
                                         format = 'BED')

xyzs <- lapply(xyz.s, function(.ele) {
  .ele <- .ele[c(18, 27, 32), -1]
  if(any(is.na(.ele))) return(NULL)
  return(.ele)
})
keep <- lengths(xyzs)>0

cd <- cellDistance(xyzs[keep], distance_method = 'DSDC')
cc <- cellClusters(cd)
cell_groups <- cutree(cc, k=3)

## check SDC (the mean of distance from each point to the centroid) for A1, B1, C1 
sdc <- sapply(xyzs[keep], SDC)
sdc <- split(sdc, cell_groups)
bp <- boxplot(sdc)

abc_ol <- which.min(bp$stats[3, ]) ## find out which group is A1, B1 overlapped
abc_nol <- which.max(bp$stats[3, ]) ## find out which group is A1, B1 not overlapped

## aggregate, this method only work for real 3D structure
#cg_xyz <- aggregateXYZs(xyz.list)
## get the cells which can represent the group
centers_cell <- getMedoidId(cd, cell_groups, N = 2)
centers_cell
cg_xyz <- xyz.s[unlist(centers_cell)]

## help function to get the coordinates in genome.
st <- function(i, offset=28000001, interval=30000){
  return(interval*i+offset)
}

## add genomic coordinates to XYZ matrix and make a GRanges object
cg_xyz <- lapply(cg_xyz, function(.ele){
  with(.ele, GRanges('chr21',
                     IRanges(st(.ele$Segment.index-1, offset=start(range)),
                             width = 30000),
                     x=X, y=Y, z=Z))
})


ABC <- GRanges('chr21', IRanges(st(c(18, 27, 32)-1, offset=start(range)), width = 30000),
               col=c('red', 'cyan', 'yellow'))
ABC$label <- c('A1', 'B1', 'C1')
ABC$type <- 'tad'
ABC$lwd <- 8
feature.gr$lwd <- 2
ABC.features <- c(feature.gr, ABC)

## visualize the grouped structures.
grouped_cell <- lapply(cg_xyz, function(cell){
  cell <- view3dStructure(cell, feature.gr=ABC.features,
                          genomicSigs = imr90_genomicSigs,
                          reverseGenomicSigs = FALSE,,
                          renderer = 'none', region = range,
                          show_coor=TRUE, lwd.backbone = 0.25,
                          lwd.gene = 0.5,
                          length.arrow = unit(0.1, "native"))
})

title <- c("A1,B1 overlap", "A1,B1 do not overlap")
showPairs(grouped_cell[[(abc_ol-1)*2+1]], grouped_cell[[(abc_nol-1)*2+1]],
          title = title,
          height = NULL)
showPairs(grouped_cell[[(abc_ol-1)*2+2]], grouped_cell[[(abc_nol-1)*2+2]],
          title = title,
          height = NULL)
```


## present single cell 3D structure with TADs based on the point cluster

The data were from [diffraction-limited experiments for IMR90 cells](https://doi.org/10.1126/science.aau1783).

```{r}
# Fig. S7
xyz <- read.csv('https://github.com/BogdanBintu/ChromatinImaging/raw/refs/heads/master/Data/IMR90_chr21-18-20Mb.csv', skip = 1)
xyz.s <- split(xyz[, -1], xyz$Chromosome.index)

#cd <- cellDistance(xyzs=lapply(xyz.s, function(.ele) fill_NA(.ele[, -1])), distance_method = 'NID',  eps = 'auto', rescale = FALSE)
cd <- readRDS(system.file('extdata', 'ChromatinImaging', 'FigS7.cellDistance.nid.rds', package = 'geomeTriD.documentation'))
cd1 <- as.matrix(cd)
cc1 <- apply(cd1[c(8, 9, 15, 26, 56), ], ## select some typical cells
             1, function(.ele){
  order(.ele)
})
cc <- cellClusters(cd)
#plot(cc, labels=FALSE)
cell_groups <- cutree(cc, h=0.5)

range_18_20 <- GRanges("chr21:18627683-20577682")
# feature.gr_18_20 <- getFeatureGR(txdb = TxDb.Hsapiens.UCSC.hg38.knownGene,
#                                  org = org.Hs.eg.db,
#                                  range = range_18_20)
## apply matlab.like2 colors for backbone
resolution <- 3
backbone_colors <- rev(matlab.like2(n=resolution*nrow(xyz.s[[1]])))

## help function to get the coordinates in genome.
st <- function(i, offset=28000001, interval=30000){
  return(interval*i+offset)
}

imr90_18_20 <- lapply(xyz.s, function(.ele){
  with(.ele, GRanges('chr21', IRanges(st(.ele$Segment.index-1,
                                         offset=start(range_18_20)),
                                      width = 30000),
                     x=X, y=Y, z=Z))
})

cells_18_20 <- list()
for(i in as.numeric(cc1[1:2, ])){
  try({
    cell <- imr90_18_20[[i]]
    cell <- cell[apply(mcols(cell), 1, function(.ele) !any(is.na(.ele)))]
    cell <- view3dStructure(cell, #feature.gr=feature.gr_18_20,
                            renderer = 'none', region = range_18_20,
                            show_coor=TRUE, lwd.backbone = 2,
                            col.backbone=backbone_colors,
                            resolution=resolution,
                            length.arrow = unit(0.1, "native"),
                            cluster3Dpoints = TRUE)
    ## reset the color for the point clusters
    pcn <- names(cell)[grepl('pointCluster', names(cell))]
    for(j in seq_along(pcn)){
      cell[[pcn[j]]]$colors <- j+1
    }
    cells_18_20[[i]] <- cell
  })
}

showPairs(cells_18_20[[cc1[1, '9']]], cells_18_20[[cc1[2, '9']]],
          height = NULL)# cell 1

showPairs(cells_18_20[[cc1[1, '56']]], cells_18_20[[cc1[2, '56']]],
          height = NULL)# cell 2

showPairs(cells_18_20[[cc1[1, '8']]], cells_18_20[[cc1[2, '8']]],
          height = NULL) # cell 3

showPairs(cells_18_20[[cc1[1, '26']]], cells_18_20[[cc1[2, '26']]],
          height = NULL)
```


## SessionInfo

```{r sessionInfo}
sessionInfo()
```