<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Figure 5C Vignette â€¢ geomeTriD.documentation</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Figure 5C Vignette">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">geomeTriD.documentation</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Fig5C.html">Figure 5C Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/geomeTriD_documentation" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Figure 5C Vignette</h1>
                        <h4 data-toc-skip class="author">Jianhong
Ou</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/geomeTriD_documentation/blob/main/vignettes/Fig5C.Rmd" class="external-link"><code>vignettes/Fig5C.Rmd</code></a></small>
      <div class="hidden name"><code>Fig5C.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="figure-5c">Figure 5C<a class="anchor" aria-label="anchor" href="#figure-5c"></a>
</h2>
<p>Figure 5C showcases the geomeTriD package, demonstrating how it
presents 3D models along with multiple genomic signals mapped onto
single-cell 3D structures.</p>
<div class="section level3">
<h3 id="load-libraries">Load Libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jianhong/geomeTriD" class="external-link">geomeTriD</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jianhong.github.io/geomeTriD_documentation">geomeTriD.documentation</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">colorRamps</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Mmusculus.UCSC.mm10.knownGene</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davidcsterratt.github.io/geometry/" class="external-link">geometry</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="present-single-cell-3d-structure-for-chair-data">Present single cell 3D structure for ChAIR data<a class="anchor" aria-label="anchor" href="#present-single-cell-3d-structure-for-chair-data"></a>
</h3>
<p>Here we present 3D genome models with integrated genomic signals
derived from <a href="https://doi.org/10.1038/s41592-025-02658-7" class="external-link">Tri-omic</a>, also
known as ChAIR (capture of chromatin Accessibility, Interaction, and RNA
profiles). This method enables simultaneous mapping of chromatin
accessibility, chromatin interactions, and gene expression. Paired-end
tag (PET) data are used to reconstruct 3D genome structures, while the
accompanying RNA-seq and ATAC-seq signals are visualized along these
models.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot region</span></span>
<span><span class="va">range</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">'chrX:1-20000000'</span><span class="op">)</span></span>
<span><span class="co">## prepare features</span></span>
<span><span class="va">feature.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getFeatureGR.html">getFeatureGR</a></span><span class="op">(</span>txdb <span class="op">=</span> <span class="va">TxDb.Mmusculus.UCSC.mm10.knownGene</span>,</span>
<span>                           org <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                           range <span class="op">=</span> <span class="va">range</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   66 genes were dropped because they have exons located on both strands of the</span></span>
<span><span class="co">##   same reference sequence or on more than one reference sequence, so cannot be</span></span>
<span><span class="co">##   represented by a single genomic range.</span></span>
<span><span class="co">##   Use 'single.strand.genes.only=FALSE' to get all the genes in a GRangesList</span></span>
<span><span class="co">##   object, or use suppressMessages() to suppress this message.</span></span></code></pre>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load the 3D structure for PET data</span></span>
<span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'G1'</span>, <span class="st">'S'</span>, <span class="st">'G2M'</span><span class="op">)</span></span>
<span><span class="va">extdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'ChAIR'</span>, package <span class="op">=</span> <span class="st">'geomeTriD.documentation'</span><span class="op">)</span></span>
<span><span class="va">ChAIR_PET_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import3dg.html">import3dg</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">'PET'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'patski.allele.'</span>, <span class="va">stage</span>, <span class="st">'_2_m.pairs.chrX.3dg.gz'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ChAIR_PET_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import3dg.html">import3dg</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">'PET'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'patski.allele.'</span>, <span class="va">stage</span>, <span class="st">'_2_p.pairs.chrX.3dg.gz'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get the chrX coordinates for maternal and paternal</span></span>
<span><span class="va">ChAIR_PET_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ChAIR_PET_m</span>, <span class="kw">function</span><span class="op">(</span><span class="va">.ele</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">.ele</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'UCSC'</span></span>
<span>  <span class="va">.ele</span><span class="op">$</span><span class="va">parental</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">.ele</span>, <span class="va">range</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="va">ChAIR_PET_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ChAIR_PET_p</span>, <span class="kw">function</span><span class="op">(</span><span class="va">.ele</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">.ele</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'UCSC'</span></span>
<span>  <span class="va">.ele</span><span class="op">$</span><span class="va">parental</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">.ele</span>, <span class="va">range</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># import the signals for ATAC-seq and RNA-seq for maternal and paternal</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">stage</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cc</span><span class="op">)</span><span class="op">{</span> <span class="co">## maternal</span></span>
<span>  <span class="fu"><a href="../reference/importGenomicSigs.html">importGenomicSigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ATAC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">'ATAC'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'patski.allele.'</span>, <span class="va">cc</span>, <span class="st">'_2.M.bw'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      RNA<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">'RNA'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'patski.allele.'</span>, <span class="va">cc</span>, <span class="st">'_2.M.bw'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      range <span class="op">=</span> <span class="va">range</span>,</span>
<span>                    cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'darkgreen'</span>, <span class="st">'darkred'</span><span class="op">)</span>,</span>
<span>                    format <span class="op">=</span> <span class="st">'BigWig'</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="va">P</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">stage</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cc</span><span class="op">)</span><span class="op">{</span> <span class="co">## paternal</span></span>
<span>  <span class="fu"><a href="../reference/importGenomicSigs.html">importGenomicSigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ATAC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">'ATAC'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'patski.allele.'</span>, <span class="va">cc</span>, <span class="st">'_2.M.bw'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      RNA<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">'RNA'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'patski.allele.'</span>, <span class="va">cc</span>, <span class="st">'_2.M.bw'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      range <span class="op">=</span> <span class="va">range</span>,</span>
<span>                    cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'darkgreen'</span>, <span class="st">'darkred'</span><span class="op">)</span>,</span>
<span>                    format <span class="op">=</span> <span class="st">'BigWig'</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## set backbone color, make it easy to catch the genomic coordinates</span></span>
<span><span class="va">resolution</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">backbone_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">resolution</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ChAIR_PET_m</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## only label the genes with high expression</span></span>
<span><span class="va">labelHighExpGene</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sig</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">this.feature.gr</span> <span class="op">&lt;-</span> <span class="va">feature.gr</span></span>
<span>  <span class="va">expressed</span> <span class="op">&lt;-</span> <span class="va">sig</span><span class="op">$</span><span class="va">RNA</span><span class="op">$</span><span class="va">dat</span><span class="op">[</span><span class="va">sig</span><span class="op">$</span><span class="va">RNA</span><span class="op">$</span><span class="va">dat</span><span class="op">$</span><span class="va">score</span><span class="op">&gt;</span><span class="fl">100</span><span class="op">]</span></span>
<span>  <span class="va">this.feature.gr</span><span class="op">$</span><span class="va">label</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">countOverlaps</a></span><span class="op">(</span><span class="va">feature.gr</span>, <span class="va">expressed</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span>  <span class="co">## remove the genes with big size</span></span>
<span>  <span class="va">this.feature.gr</span><span class="op">$</span><span class="va">label</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">this.feature.gr</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">100000</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span>  <span class="co">## only show the first gene label if multiple high expressed gene</span></span>
<span>  <span class="co">## in the same region</span></span>
<span>  <span class="va">l.rle</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rle.html" class="external-link">rle</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">this.feature.gr</span><span class="op">$</span><span class="va">label</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">l.rle</span><span class="op">$</span><span class="va">values</span><span class="op">[</span><span class="va">l.rle</span><span class="op">$</span><span class="va">lengths</span><span class="op">&lt;=</span><span class="fl">5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span> <span class="co">## merge the region with small gap</span></span>
<span>  <span class="va">l.rle</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rle.html" class="external-link">rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rle.html" class="external-link">inverse.rle</a></span><span class="op">(</span><span class="va">l.rle</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">l.rle</span><span class="op">$</span><span class="va">lengths</span><span class="op">)</span></span>
<span>  <span class="va">this.feature.gr</span><span class="op">$</span><span class="va">label</span><span class="op">[</span><span class="op">-</span><span class="va">l</span><span class="op">[</span><span class="va">l.rle</span><span class="op">$</span><span class="va">values</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">this.feature.gr</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## create threeJsGeometry objects for maternal and paternal cells</span></span>
<span><span class="va">M_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">ChAIR_PET_m</span>, <span class="va">M</span>, FUN<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">model</span>, <span class="va">sig</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geomeTriD/man/view3dStructure.html" class="external-link">view3dStructure</a></span><span class="op">(</span><span class="va">model</span>,</span>
<span>                  feature.gr <span class="op">=</span> <span class="fu">labelHighExpGene</span><span class="op">(</span><span class="va">sig</span><span class="op">)</span>,</span>
<span>                  genomicSigs <span class="op">=</span> <span class="va">sig</span>,</span>
<span>                  signalTransformFun <span class="op">=</span> <span class="va">log1p</span>,</span>
<span>                  lwd.maxGenomicSigs <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                  reverseGenomicSigs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                  renderer <span class="op">=</span> <span class="st">'none'</span>,</span>
<span>                  resolution <span class="op">=</span> <span class="va">resolution</span>,</span>
<span>                  col.backbone <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">resolution</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  lwd.backbone <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                  lwd.gene <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  region <span class="op">=</span> <span class="va">range</span>,</span>
<span>                  show_coor <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                  cluster3Dpoints <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span>, SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">P_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">ChAIR_PET_p</span>, <span class="va">P</span>, FUN<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">model</span>, <span class="va">sig</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geomeTriD/man/view3dStructure.html" class="external-link">view3dStructure</a></span><span class="op">(</span><span class="va">model</span>,</span>
<span>                  feature.gr <span class="op">=</span> <span class="fu">labelHighExpGene</span><span class="op">(</span><span class="va">sig</span><span class="op">)</span>,</span>
<span>                  genomicSigs <span class="op">=</span> <span class="va">sig</span>,</span>
<span>                  signalTransformFun <span class="op">=</span> <span class="va">log1p</span>,</span>
<span>                  lwd.maxGenomicSigs <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                  reverseGenomicSigs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                  renderer <span class="op">=</span> <span class="st">'none'</span>,</span>
<span>                  resolution <span class="op">=</span> <span class="va">resolution</span>,</span>
<span>                  col.backbone <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">resolution</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  lwd.backbone <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                  lwd.gene <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  region <span class="op">=</span> <span class="va">range</span>,</span>
<span>                  show_coor <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                  cluster3Dpoints <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span>, SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## show the maternal and paternal structures side by side</span></span>
<span><span class="fu"><a href="../reference/showPairs.html">showPairs</a></span><span class="op">(</span><span class="va">M_cells</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">P_cells</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, title<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mat G1'</span>, <span class="st">'pat G1'</span><span class="op">)</span><span class="op">)</span> <span class="co">#G1</span></span>
<span><span class="fu"><a href="../reference/showPairs.html">showPairs</a></span><span class="op">(</span><span class="va">M_cells</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">P_cells</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, title<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mat S'</span>, <span class="st">'pat S'</span><span class="op">)</span><span class="op">)</span> <span class="co">#S</span></span>
<span><span class="fu"><a href="../reference/showPairs.html">showPairs</a></span><span class="op">(</span><span class="va">M_cells</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, <span class="va">P_cells</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, title<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mat G2M'</span>, <span class="st">'pat G2M'</span><span class="op">)</span><span class="op">)</span> <span class="co">#G2M</span></span>
<span></span>
<span><span class="co">## check the size </span></span>
<span><span class="va">getV</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">points</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">vol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geometry/man/convhulln.html" class="external-link">convhulln</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">points</span><span class="op">)</span><span class="op">)</span>, options<span class="op">=</span><span class="st">'Fa'</span><span class="op">)</span><span class="op">$</span><span class="va">vol</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ChAIR_PET_m</span>, <span class="va">getV</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ChAIR_PET_p</span>, <span class="va">getV</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] geometry_0.5.2                           </span></span>
<span><span class="co">##  [2] org.Mm.eg.db_3.21.0                      </span></span>
<span><span class="co">##  [3] TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0</span></span>
<span><span class="co">##  [4] GenomicFeatures_1.61.4                   </span></span>
<span><span class="co">##  [5] AnnotationDbi_1.71.0                     </span></span>
<span><span class="co">##  [6] Biobase_2.69.0                           </span></span>
<span><span class="co">##  [7] colorRamps_2.3.4                         </span></span>
<span><span class="co">##  [8] GenomicRanges_1.61.1                     </span></span>
<span><span class="co">##  [9] Seqinfo_0.99.1                           </span></span>
<span><span class="co">## [10] IRanges_2.43.0                           </span></span>
<span><span class="co">## [11] S4Vectors_0.47.0                         </span></span>
<span><span class="co">## [12] BiocGenerics_0.55.0                      </span></span>
<span><span class="co">## [13] generics_0.1.4                           </span></span>
<span><span class="co">## [14] geomeTriD.documentation_0.0.4            </span></span>
<span><span class="co">## [15] geomeTriD_1.3.14                         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] BiocIO_1.19.0               bitops_1.0-9               </span></span>
<span><span class="co">##   [3] filelock_1.0.3              tibble_3.3.0               </span></span>
<span><span class="co">##   [5] R.oo_1.27.1                 XML_3.99-0.18              </span></span>
<span><span class="co">##   [7] rpart_4.1.24                lifecycle_1.0.4            </span></span>
<span><span class="co">##   [9] httr2_1.1.2                 aricode_1.0.3              </span></span>
<span><span class="co">##  [11] globals_0.18.0              lattice_0.22-7             </span></span>
<span><span class="co">##  [13] ensembldb_2.33.1            MASS_7.3-65                </span></span>
<span><span class="co">##  [15] backports_1.5.0             magrittr_2.0.3             </span></span>
<span><span class="co">##  [17] Hmisc_5.2-3                 sass_0.4.10                </span></span>
<span><span class="co">##  [19] rmarkdown_2.29              jquerylib_0.1.4            </span></span>
<span><span class="co">##  [21] yaml_2.3.10                 plotrix_3.8-4              </span></span>
<span><span class="co">##  [23] Gviz_1.53.1                 DBI_1.2.3                  </span></span>
<span><span class="co">##  [25] RColorBrewer_1.1-3          abind_1.4-8                </span></span>
<span><span class="co">##  [27] R.utils_2.13.0              AnnotationFilter_1.33.0    </span></span>
<span><span class="co">##  [29] biovizBase_1.57.1           RCurl_1.98-1.17            </span></span>
<span><span class="co">##  [31] rgl_1.3.24                  nnet_7.3-20                </span></span>
<span><span class="co">##  [33] VariantAnnotation_1.55.1    rappdirs_0.3.3             </span></span>
<span><span class="co">##  [35] grImport_0.9-7              listenv_0.9.1              </span></span>
<span><span class="co">##  [37] parallelly_1.45.0           pkgdown_2.1.3              </span></span>
<span><span class="co">##  [39] codetools_0.2-20            DelayedArray_0.35.2        </span></span>
<span><span class="co">##  [41] xml2_1.3.8                  tidyselect_1.2.1           </span></span>
<span><span class="co">##  [43] UCSC.utils_1.5.0            farver_2.1.2               </span></span>
<span><span class="co">##  [45] matrixStats_1.5.0           BiocFileCache_2.99.5       </span></span>
<span><span class="co">##  [47] base64enc_0.1-3             GenomicAlignments_1.45.1   </span></span>
<span><span class="co">##  [49] jsonlite_2.0.0              trackViewer_1.45.1         </span></span>
<span><span class="co">##  [51] progressr_0.15.1            Formula_1.2-5              </span></span>
<span><span class="co">##  [53] systemfonts_1.2.3           dbscan_1.2.2               </span></span>
<span><span class="co">##  [55] tools_4.5.1                 progress_1.2.3             </span></span>
<span><span class="co">##  [57] ragg_1.4.0                  strawr_0.0.92              </span></span>
<span><span class="co">##  [59] Rcpp_1.0.14                 glue_1.8.0                 </span></span>
<span><span class="co">##  [61] gridExtra_2.3               SparseArray_1.9.0          </span></span>
<span><span class="co">##  [63] xfun_0.52                   MatrixGenerics_1.21.0      </span></span>
<span><span class="co">##  [65] GenomeInfoDb_1.45.6         dplyr_1.1.4                </span></span>
<span><span class="co">##  [67] fastmap_1.2.0               latticeExtra_0.6-30        </span></span>
<span><span class="co">##  [69] rhdf5filters_1.21.0         digest_0.6.37              </span></span>
<span><span class="co">##  [71] R6_2.6.1                    textshaping_1.0.1          </span></span>
<span><span class="co">##  [73] colorspace_2.1-1            jpeg_0.1-11                </span></span>
<span><span class="co">##  [75] dichromat_2.0-0.1           biomaRt_2.65.0             </span></span>
<span><span class="co">##  [77] RSQLite_2.4.1               R.methodsS3_1.8.2          </span></span>
<span><span class="co">##  [79] data.table_1.17.6           rtracklayer_1.69.1         </span></span>
<span><span class="co">##  [81] prettyunits_1.2.0           InteractionSet_1.37.0      </span></span>
<span><span class="co">##  [83] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">##  [85] S4Arrays_1.9.1              pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [87] gtable_0.3.6                blob_1.2.4                 </span></span>
<span><span class="co">##  [89] XVector_0.49.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">##  [91] ProtGenerics_1.41.0         clue_0.3-66                </span></span>
<span><span class="co">##  [93] scales_1.4.0                png_0.1-8                  </span></span>
<span><span class="co">##  [95] knitr_1.50                  rstudioapi_0.17.1          </span></span>
<span><span class="co">##  [97] rjson_0.2.23                checkmate_2.3.2            </span></span>
<span><span class="co">##  [99] magic_1.6-1                 curl_6.4.0                 </span></span>
<span><span class="co">## [101] cachem_1.1.0                rhdf5_2.53.1               </span></span>
<span><span class="co">## [103] stringr_1.5.1               parallel_4.5.1             </span></span>
<span><span class="co">## [105] foreign_0.8-90              restfulr_0.0.16            </span></span>
<span><span class="co">## [107] desc_1.4.3                  pillar_1.10.2              </span></span>
<span><span class="co">## [109] vctrs_0.6.5                 RANN_2.6.2                 </span></span>
<span><span class="co">## [111] dbplyr_2.5.0                cluster_2.1.8.1            </span></span>
<span><span class="co">## [113] htmlTable_2.4.3             evaluate_1.0.4             </span></span>
<span><span class="co">## [115] cli_3.6.5                   compiler_4.5.1             </span></span>
<span><span class="co">## [117] Rsamtools_2.25.1            rlang_1.1.6                </span></span>
<span><span class="co">## [119] crayon_1.5.3                future.apply_1.20.0        </span></span>
<span><span class="co">## [121] interp_1.1-6                fs_1.6.6                   </span></span>
<span><span class="co">## [123] stringi_1.8.7               deldir_2.0-4               </span></span>
<span><span class="co">## [125] BiocParallel_1.43.4         txdbmaker_1.5.6            </span></span>
<span><span class="co">## [127] Biostrings_2.77.2           lazyeval_0.2.2             </span></span>
<span><span class="co">## [129] Matrix_1.7-3                BSgenome_1.77.1            </span></span>
<span><span class="co">## [131] hms_1.1.3                   bit64_4.6.0-1              </span></span>
<span><span class="co">## [133] future_1.58.0               ggplot2_3.5.2              </span></span>
<span><span class="co">## [135] Rhdf5lib_1.31.0             KEGGREST_1.49.1            </span></span>
<span><span class="co">## [137] SummarizedExperiment_1.39.1 igraph_2.1.4               </span></span>
<span><span class="co">## [139] memoise_2.0.1               bslib_0.9.0                </span></span>
<span><span class="co">## [141] bit_4.6.0</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
